### DISCUSSION TITLE
Windows service doesn't use "RABBITMQ_CONFIG_FILE" environment variable

### DISCUSSION THREAD

**Discussion by geniys** [rabbitmq/rabbitmq-server#3790] [Questions]
In RabbitMQ 3.8.3, the "RABBITMQ_CONFIG_FILE" environment variable worked when installing as a service.
But it doesn't work with RabbitMQ 3.9.10. It only works when run it with "rabbitmq-server.bat".

My environments is
Erlan OTP : 23.3.4.9
RabbitMQ version : 3.9.10
Platform : Windows Server 2019

Reproduce steps
```
SET ERLANG_HOME=C:\Program Files\erl-23.3.4.9
SET RABBITMQ_BASE=C:\Works\rabbitmq\data
SET RABBITMQ_CONFIG_FILE=C:\Works\rabbitmq\conf\rabbitmq
SET RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS=+Q 75000

rabbitmq-service.bat install 
rabbitmq-service.bat start
```
then log shows as "**config file(s) : (none)**"
```
2021-11-23 16:14:06.219000+09:00 [info] <0.219.0>  Starting RabbitMQ 3.9.10 on Erlang 23.3.4.9 [emu]
2021-11-23 16:14:06.219000+09:00 [info] <0.219.0>  Copyright (c) 2007-2021 VMware, Inc. or its affiliates.
2021-11-23 16:14:06.219000+09:00 [info] <0.219.0>  Licensed under the MPL 2.0. Website: https://rabbitmq.com
2021-11-23 16:14:06.234000+09:00 [info] <0.219.0> 
2021-11-23 16:14:06.234000+09:00 [info] <0.219.0>  node           : rabbit@WIN-7MH4DE06VDS
2021-11-23 16:14:06.234000+09:00 [info] <0.219.0>  home dir       : c:/Windows/system32/config/systemprofile
2021-11-23 16:14:06.234000+09:00 [info] <0.219.0>  config file(s) : (none)
2021-11-23 16:14:06.234000+09:00 [info] <0.219.0>  cookie hash    : hdxOFDdykEqWA/sCV8zJZg==
2021-11-23 16:14:06.234000+09:00 [info] <0.219.0>  log(s)         : <stdout>
2021-11-23 16:14:06.234000+09:00 [info] <0.219.0>                 : c:/Works/rabbitmq/data/log/rabbit@WIN-7MH4DE06VDS.log
2021-11-23 16:14:06.234000+09:00 [info] <0.219.0>                 : c:/Works/rabbitmq/data/log/rabbit@WIN-7MH4DE06VDS_upgrade.log
2021-11-23 16:14:06.234000+09:00 [info] <0.219.0>  database dir   : c:/Works/rabbitmq/data/db/rabbit@WIN-7MH4DE06VDS-mnesia
```

But run with "rabbitmq-server.bat", then log shows as
```
2021-11-23 16:11:25.811000+09:00 [info] <0.222.0>  Starting RabbitMQ 3.9.10 on Erlang 23.3.4.9 [emu]
2021-11-23 16:11:25.811000+09:00 [info] <0.222.0>  Copyright (c) 2007-2021 VMware, Inc. or its affiliates.
2021-11-23 16:11:25.811000+09:00 [info] <0.222.0>  Licensed under the MPL 2.0. Website: https://rabbitmq.com
2021-11-23 16:11:25.811000+09:00 [info] <0.222.0> 
2021-11-23 16:11:25.811000+09:00 [info] <0.222.0>  node           : rabbit@WIN-7MH4DE06VDS
2021-11-23 16:11:25.811000+09:00 [info] <0.222.0>  home dir       : c:/Users/dev
2021-11-23 16:11:25.811000+09:00 [info] <0.222.0>  config file(s) : c:/Works/rabbitmq/conf/rabbitmq.conf
2021-11-23 16:11:25.811000+09:00 [info] <0.222.0>  cookie hash    : 4Y+9yh7QrOGXO8aIm4YQZw==
2021-11-23 16:11:25.811000+09:00 [info] <0.222.0>  log(s)         : <stdout>
2021-11-23 16:11:25.811000+09:00 [info] <0.222.0>                 : c:/Works/rabbitmq/data/log/rabbit@WIN-7MH4DE06VDS.log
2021-11-23 16:11:25.811000+09:00 [info] <0.222.0>                 : c:/Works/rabbitmq/data/log/rabbit@WIN-7MH4DE06VDS_upgrade.log
2021-11-23 16:11:25.811000+09:00 [info] <0.222.0>  database dir   : c:/Works/rabbitmq/data/db/rabbit@WIN-7MH4DE06VDS-mnesia
```

How can I change the config file path when installing as a windows service?

**Comment by geniys** [comment]
Now I realized that if I register the environment variable in the system's persistent area, it works.
For example,
```
setx -m RABBITMQ_CONFIG_FILE "C:\Works\rabbitmq\conf\rabbitmq.conf"
```

Is the current behavior as expected?

I think it is better to capture and store the environment variables to service property when installing the service because it is safe from unexpected modifications.

**Comment by lukebakken** [comment]
This behavior changed between version 3.8.3 and 3.8.4. I'm investigating the commit logs to see why.

**Reply by lukebakken** [reply]
https://github.com/rabbitmq/rabbitmq-server/pull/2142

https://github.com/rabbitmq/rabbitmq-server/pull/2282

This changed as the result of a major refactor in how RabbitMQ starts up. You have found one workaround. I believe there is another way to specify the config file that is "baked into" the service registration. I'll look into that.

**Reply by lukebakken** [reply]
OK, as of the current version of RabbitMQ the only way to set `RABBITMQ_CONFIG_FILE` for the RabbitMQ windows service is to set a system-wide environment variable as @geniys describes here -

https://github.com/rabbitmq/rabbitmq-server/discussions/3790#discussioncomment-1686014

@michaelklishin @dumbbell there were several env variables that used to be "baked into" the Windows Service registry settings:

https://github.com/rabbitmq/rabbitmq-server/pull/2142/files#diff-0d61bdfcf7328b3eb12e30df2e912e4b0eea0f60dcb47ccc53353c20c71d36f6L319-L331

Basically anything that can be set in the Windows environment could be passed to `erlsrv.exe` via the `-env` argument. Do you think it's worth re-visiting those `RABBITMQ_*` environment variables that used to be part of the service definition for re-inclusion into `rabbitmq-service.bat`?

Note that setting these variables as system-wide env variables is a functional workaround.
